README.md
# Date Hover Preview

## Features

Hover over a string in any document that represents a date, and get a small preview that converts that date to a configurable set of timezones and formats.

Hoverable formats currently include:

* [Unix Time](https://en.wikipedia.org/wiki/Unix_time) (9-10 digits as seconds/11-14 digits as milliseconds)
* [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601)

![Preview](resources/readme/dateHoverPreviewExample.gif)

## Extension Settings

* `date-hover-preview.detect.ISO-8601String`: Whether to detect [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) format date string when hovering.
* `date-hover-preview.detect.unixTime`: Whether to detect Unix epoch time when hovering. Currently supports 9-10 digits (as seconds) or up to 14 digits (as milliseconds).
* `date-hover-preview.primaryPreview.enable`: Show primary date preview.
* `date-hover-preview.primaryPreview.name`: Show as the name of primary preview section.
* `date-hover-preview.primaryPreview.format`: Primary date preview format, described in [day.js format documentation](https://day.js.org/docs/en/display/format). Leave empty to use ISO-8601 format. See below for a quick reference.
* `date-hover-preview.primaryPreview.utcOffset`: UTC offset for the primary preview expressed as a number of hours between -16 and 16 (any number outside of that range will be interpreted as minutes).
* `date-hover-preview.primaryPreview.timezone`: Timezone for the primary preview. Timezones are expressed like \"PST\" or \"America/Los_Angeles\". This setting will take precedence over `date-hover-preview.primaryPreview.utcOffset`. If both are left blank this will use your local timezone. For a list of timezones, see the [TZ database name column](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones).
* `date-hover-preview.additionalPreviews`: An array of additional preview items.
* `date-hover-preview.additionalPreviews[].name`: Show as the name of this additional preview.
* `date-hover-preview.additionalPreviews[].format`: Used to format preview display. Leave empty to use ISO-8601 format.
* `date-hover-preview.additionalPreviews[].utcOffset`: UTC offset expressed as a number of hours between -16 and 16 (any number outside of that range will be interpreted as minutes).
* `date-hover-preview.additionalPreviews[].timezone`: Timezones are expressed like \"PST\" or \"America/Los_Angeles\". This setting will take precedence over utcOffset. If both are left blank this will use your local timezone.

### Format Quick Reference

| Format | Output           | Description                           |
| ------ | ---------------- | ------------------------------------- |
| `YY`   | 18               | Two-digit year                        |
| `YYYY` | 2018             | Four-digit year                       |
| `M`    | 1-12             | The month, beginning at 1             |
| `MM`   | 01-12            | The month, 2-digits                   |
| `MMM`  | Jan-Dec          | The abbreviated month name            |
| `MMMM` | January-December | The full month name                   |
| `D`    | 1-31             | The day of the month                  |
| `DD`   | 01-31            | The day of the month, 2-digits        |
| `d`    | 0-6              | The day of the week, with Sunday as 0 |
| `dd`   | Su-Sa            | The min name of the day of the week   |
| `ddd`  | Sun-Sat          | The short name of the day of the week |
| `dddd` | Sunday-Saturday  | The name of the day of the week       |
| `H`    | 0-23             | The hour                              |
| `HH`   | 00-23            | The hour, 2-digits                    |
| `h`    | 1-12             | The hour, 12-hour clock               |
| `hh`   | 01-12            | The hour, 12-hour clock, 2-digits     |
| `m`    | 0-59             | The minute                            |
| `mm`   | 00-59            | The minute, 2-digits                  |
| `s`    | 0-59             | The second                            |
| `ss`   | 00-59            | The second, 2-digits                  |
| `SSS`  | 000-999          | The millisecond, 3-digits             |
| `Z`    | +05:00           | The offset from UTC, ±HH:mm           |
| `ZZ`   | +0500            | The offset from UTC, ±HHmm            |
| `A`    | AM PM            |                                       |
| `a`    | am pm            |                                       |


## Upcoming Features

Customizable detection formats

## Known Issues

None

CHANGELOG.md
# Change Log

### [1.0.0] - 2021/2/26

- Initial release

### [1.0.3] - 2021/3/1

- Fix: Changed the activation event because sometimes the extension would not become active


src/DateHoverProvider.ts
import * as vscode from 'vscode';
import * as dayjs from 'dayjs';
import * as utcPlugin from 'dayjs/plugin/utc';
import * as timezonePlugin from 'dayjs/plugin/timezone';

dayjs.extend(utcPlugin);
dayjs.extend(timezonePlugin);

class DateHoverProvider implements vscode.HoverProvider {

    private logger = vscode.window.createOutputChannel("DateHoverPreview");

    public constructor() {
        this.log("DateHoverProvider activated")
    }

    private dateRegex = {
        iso8601: /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{3})?(Z|(\+|\-)\d{2}:\d{2})/,
        epoch: /\d{9,14}/
    };

    get detectIso8601Enabled(): boolean {
        return vscode.workspace.getConfiguration().get<boolean>('date-hover-preview.detect.ISO-8601String', true);
    }

    get detectEpochEnabled(): boolean {
        return vscode.workspace.getConfiguration().get<boolean>('date-hover-preview.detect.epochTimestamp', true);
    }
    
    get primaryPreviewEnabled(): boolean {
        return vscode.workspace.getConfiguration().get<boolean>('date-hover-preview.primaryPreview.enable', true);
    }

    get primaryPreviewName(): string {
        return vscode.workspace.getConfiguration().get<string>('date-hover-preview.primaryPreview.name', 'Local');
    }

    get primaryPreviewFormat(): string {
        return vscode.workspace.getConfiguration().get<string>('date-hover-preview.primaryPreview.format', '');
    }
    
    get primaryPreviewUtcOffset(): number | undefined {
        return vscode.workspace.getConfiguration().get('date-hover-preview.primaryPreview.utcOffset');
    }

    get primaryPreviewTimezone(): string | undefined {
        return vscode.workspace.getConfiguration().get<string>('date-hover-preview.primaryPreview.timezone');
    }

    get additionalPreivews(): Array<PreviewConfig> {
        return vscode.workspace.getConfiguration().get<Array<PreviewConfig>>('date-hover-preview.additionalPreviews', []);
    }

    provideHover(document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken): vscode.ProviderResult<vscode.Hover> {
        
      let date: any = undefined;

        if (this.detectIso8601Enabled) {
            const isoRange = document.getWordRangeAtPosition(position, this.dateRegex.iso8601);

            if (isoRange) {
                const hoveredWord = document.getText(isoRange);
                date = dayjs(hoveredWord);
            }
        }

        if (this.detectEpochEnabled) {
            const epochRange = document.getWordRangeAtPosition(position, this.dateRegex.epoch);

            if (epochRange) {
                let hoveredWord = document.getText(epochRange);
                // convert seconds to milliseconds
                if (hoveredWord.length <= 10) {
                    hoveredWord += '000';
                }

                date = dayjs(+hoveredWord);
            }
        }

        let previewMessage = '';
        if (date !== undefined && (date as dayjs.Dayjs).isValid()) {
            this.log("Detected date, providing hover preview");
            previewMessage = this.buildPreviewMessage(date);
        }

        return previewMessage ? new vscode.Hover(previewMessage) : undefined;
    }

    private buildPreviewMessage(date: dayjs.Dayjs): string {
        
        const previewConfigs = [...this.additionalPreivews];

        if (this.primaryPreviewEnabled) {
            const primaryPreviewConfig = {
                name: this.primaryPreviewName,
                format: this.primaryPreviewFormat,
                utcOffset: this.primaryPreviewUtcOffset,
                timezone: this.primaryPreviewTimezone
            };
            previewConfigs.unshift(primaryPreviewConfig);
        }

        let previewMessage = '';

        for (const item of previewConfigs) {

            if (!item.name) {
                continue;
            }
            
            let dateInTimezone: dayjs.Dayjs;
            if(item.timezone) {
                dateInTimezone = date.tz(item.timezone);
            } else if(item.utcOffset != undefined) {
                dateInTimezone = date.utcOffset(item.utcOffset);
            } else {
                dateInTimezone = date.tz(dayjs.tz.guess());
            }

            const formattedDate = `${item.format ? dateInTimezone.format(item.format) : dateInTimezone.format()}`;
            previewMessage += this.buildSinglePreviewItem(item.name, dateInTimezone.format('Z'), formattedDate);
        }

        return previewMessage;
    }

    private buildSinglePreviewItem(name: string, utcOffsetStr: string, dateString: string): string {
        return `\n\n*${name} (${utcOffsetStr})*    \n${dateString}`;
    }

    private log(string: string) {
        this.logger.appendLine(string);
    }
}

export default DateHoverProvider;